vault/request_handling_list_filtering.go

```diff
@@ -0,0 +1,153 @@
+// Copyright (c) 2025 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/openbao/openbao/helper/identity"
+	"github.com/openbao/openbao/sdk/v2/helper/template"
+	"github.com/openbao/openbao/sdk/v2/logical"
+)
+
+func (c *Core) filterListResponse(ctx context.Context, req *logical.Request, unauth bool, auth *logical.Auth, acl *ACL, te *logical.TokenEntry, entity *identity.Entity, resp *logical.Response) error {
+	// Ignore lists with no results.
+	if resp == nil {
+		return nil
+	}
+
+	// Ignore non-list operations.
+	switch req.Operation {
+	case logical.ListOperation:
+	case logical.ScanOperation:
+	default:
+		return nil
+	}
+
+	// No filtering, so nothing to do.
+	if auth.ResponseKeysFilterPath == "" {
+		return nil
+	}
+
+	// Secret and Auth should not be set on List operation responses which
+	// expect to be filtered.
+	if resp.Secret != nil {
+		c.logger.Error("non-empty secret on filtered list response", "path", req.Path)
+		return ErrInternalError
+	}
+	if resp.Auth != nil {
+		c.logger.Error("non-empty secret on filtered list response", "path", req.Path)
+		return ErrInternalError
+	}
+
+	// Exit early when there is no data.
+	if resp.Data == nil {
+		return nil
+	}
+
+	// Validate we have all required data fields and no unexpected ones.
+	keysRaw, present := resp.Data["keys"]
+	keyInfoRaw, keyInfoPresent := resp.Data["key_info"]
+	if !present {
+		c.logger.Error("missing required parameter keys on filtered list response", "path", req.Path)
+		return ErrInternalError
+	}
+	for keyName := range resp.Data {
+		if keyName == "keys" || keyName == "key_info" {
+			continue
+		}
+
+		c.logger.Error("unknown parameter on filtered list response", "path", req.Path, "field", keyName)
+		return ErrInternalError
+	}
+	keys, ok := keysRaw.([]string)
+	if !ok {
+		c.logger.Error("invalid type for parameter keys on filtered list response", "path", req.Path, "type", fmt.Sprintf("%T", keysRaw))
+		return ErrInternalError
+	}
+	var keyInfo map[string]interface{}
+	if keyInfoPresent {
+		keyInfo, ok = keyInfoRaw.(map[string]interface{})
+		if !ok {
+			c.logger.Error("invalid type for parameter key_info on filtered list response", "path", req.Path, "type", fmt.Sprintf("%T", keysRaw))
+			return ErrInternalError
+		}
+	}
+
+	filteredKeys := make([]string, 0, len(keys))
+	tmpl, err := compileTemplatePathForFiltering(auth.ResponseKeysFilterPath)
```

Comment:
I ended up adjust things to take templates already, as I realized we need them to support path capability grants containing wildcards.